_blog/2018/egress-https.md

```diff
@@ -0,0 +1,173 @@
+```
+
+and see our new Egress Rule in the output:
+```bash
+NAME		KIND					NAMESPACE
+googleapis	EgressRule.v1alpha2.config.istio.io	default
+```
+
+We can delete our Egress Rule:
+```bash
+istioctl delete egressrule googleapis -n default
+```
+
+and see in the output of _istioctl delete_ that the Egress Rule is deleted:
+```
+Deleted config: egressrule googleapis
+```
+
+Accessing the web page now will produce the same error that we experienced before, namely _Error fetching product details_. As we can see, the Egress Rules are defined **dynamically**, as many other Istio configuration artifacts. The Istio Operators can decide dynamically which domains they allow the microservices to access. They can enable and disable traffic to the external domains on the fly, without redeploying the microservices.
+
+## Issues with Istio Egress Traffic Control
+### TLS Origination by Istio
+There is a caveat to the story. In HTTPS, all the HTTP details (hostname, path, headers etc.) are encrypted, so Istio cannot know the destination domain of the encrypted requests. Well, Istio could know the destination domain by the  [SNI](https://tools.ietf.org/html/rfc3546#section-3.1) (_Server Name Indication_) field. This feature is not yet implemented in Istio. Therefore, currently Istio cannot perform filtering of HTTPS requests based on the destination domains.
+
+To allow Istio to perform filtering of egress requests based on domains, the microservices must issue HTTP requests. Istio then will open HTTPS connection to the destination (perform TLS origination). The code of the microservices must be written differently or configured differently, according to whether the microservice runs inside or outside an Istio Service Mesh. This contradicts the Istio Design Goal of [Maximizing Transparency]({{home}}/docs/concepts/what-is-istio/goals.html). Tough luck, sometimes we must compromise...
+
+The diagram below shows how the HTTPS traffic to external services is performed. On the top, a microservice outside an Istio Service Mesh
+sends regular HTTPS requests, encrypted end-to-end. On the bottom, the same microservice inside an Istio Service Mesh must send unencrypted HTTP requests inside a pod, which are intercepted by the sidecar Envoy proxy. The sidecar proxy performs TLS origination, so the traffic between the pod and the external service is encrypted.
+
+<figure><img src="img/https_from_the_app.svg" alt="HTTPS traffic to external services, from outside vs. from inside an Istio Service Mesh" title="HTTPS traffic to external services, from outside vs. from inside an Istio Service Mesh" />
+<figcaption>HTTPS traffic to external services, from outside vs. from inside an Istio Service Mesh</figcaption></figure>
+
+Here is how we code this behavior in the [the Bookinfo details microservice code](https://github.com/istio/istio/blob/master/samples/bookinfo/src/details/details.rb), using Ruby [net/http module](https://docs.ruby-lang.org/en/2.0.0/Net/HTTP.html):
+```ruby
```

Comment:
The ruby details take a lot of explaining...
I think we should use the curl example here of accessing external site.
`curl http://myservice.com:443/`  because it is more accessible.

On a related note, I use debian egress rule so pods can get new software ...
```
apiVersion: "config.istio.io/v1alpha2"
kind: EgressRule
metadata:
  name: debian-egress-rule
spec:
  destination:
    service: "*.debian.org"
  ports:
    - port: 80
      protocol: http
    - port: 443
      protocol: https
```

Reply:
@mandarjog I wanted to show specifically how this can be done in code: you add some environment variable, e.g. `WITH_ISTIO` and depending on it you set schema to be `http` or `https`, or set some flag as in case of `net/http` in Ruby. Then you set this environment variable in kubernetes Deployment.